### ISSUE TITLE
Release automation and keeping up project momentum

### ISSUE THREAD

**Issue by DavidSpek** [distribution/distribution#4025]
## Background
It's no secret that the project hasn't been the most active in terms of releases. It's my opinion that the slow and unpredictable release cycle is also partly responsible for the slow momentum of the project and contributions. To solve this, I would like to propose a fully automated release cycle so that contributors know when their improvements will be included in a release and maintainers don't need to manually plan and execute each release.

## The setup
1. We enforce PR titles to be named in accordance to the [Conventional Commit](https://www.conventionalcommits.org/en/v1.0.0/) specification using [this](https://github.com/amannn/action-semantic-pull-request) GitHub Action.
2. We configure the repository to only allow Squash commits and `Default to pull request title` so that the commits merged into the repo follow the the [Conventional Commit](https://www.conventionalcommits.org/en/v1.0.0/) specification. See below for the setting for the repo.
3. We configure [Semantic Release](https://github.com/semantic-release/semantic-release) to use the [Conventional Commit](https://www.conventionalcommits.org/en/v1.0.0/) specification and run it monthly using [this](https://github.com/cycjimmy/semantic-release-action) GitHub Action. Semantic Release will take care of creating the appropriate Git tags for the commits that were pushed since the last release. It also creates a release on GitHub with the changelog, and will comment on merged PRs in which release the changes have been included.

This allows us to publish new versions of `distribution/distribution` in a fully automated fashion with the release versions conforming to proper SemVer and no need for manual intervention.

***Repo setting for only allow squash merges:***
![image](https://github.com/distribution/distribution/assets/28541758/ed7da007-80ea-4e24-9a46-65decb23b0ed)

## Open points for discussion

Naturally a change like this brings up a lot of discussion points. Here are a few that I'd like to bring up. However, none of these should be blocking for the initial implementation for the automation.

### Release cadence

Having a high release cadence can be a chore for users of the software. Releasing quarterly I feel might not keep the project momentum as high as releasing monthly, but it is also a good option instead of monthly. While it would be possible to release a new version every day, week or for each new commit, since this project is used by big services as part of their infrastructure I don't believe this is a good choice.

### Branching structure

The choice of how we want to create release branches for back porting fixes has a big influence on how our PR workflow will look and how Semantic Release needs to be configured. For some background information I highly recommend reading [this](https://semantic-release.gitbook.io/semantic-release/usage/workflow-configuration) documentation for Semantic Release.

Currently we are creating branches for minor versions (e.g. `release/2.7`, `release/2.8`). When following proper SemVer, minor versions shouldn't introduce any breaking changes. As such, I'm not sure we would need to create branches for minor versions going forward (this should also reduce the maintenance burden somewhat, also taking errors with Semantic Release into account). Personally I think we should only create branches for new **major** versions.

This also brings up the discussion about what the default release branch should be. Taking the upcoming `v3` release as an example I think there are 2 setups to consider:
1. The `main` branch is and remains the release branch for `v3` going forward. This means contributors can simply target the main branch to add their improvements for the current release. This also mean we don't need to setup a bunch of automation to merge changes from main into release the branch(es). For any breaking changes, they would target the branch named `next`. Before merging `next` into `main` for the `v4` release, a branch for `v3` is created for continued support and fixes.
2. For the `v3` release a `v3` release branch is created. All contributors wanted to add improvements to the current release would need to target that branch. The `main` branch would then again be used for the next major release (similar to `next` from option 1). We'd need to setup automation for merging any minor and patch commits from the `main` branch into the `v3` release branch.

To avoid the current situation where the `main` branch has diverged enormously from the current release branch (which in my opinion seriously hinders the momentum of the project), my preference is for the first option. This should also significantly reduce the burden for maintainers, remove the need for additional complex automation workflows and be more transparent for contributors.

**Comment by DavidSpek** [comment]
CC'ing the reviewers and maintainers: @milosgajdos @chrispat @clarkbw @corhere @deleteriousEffect @heww @joaodrp @justincormack @squizzi @sargun @wy65701436 @stevelasker @dmcgowan @stevvooe @thajeztah @Jamstah

**Comment by DavidSpek** [comment]
Having a closer look at some of the semantic release documentation [here](https://github.com/semantic-release/semantic-release/blob/master/docs/recipes/release-workflow/maintenance-releases.md), it actually confirms what I thought we could do going forward. We can create the branches for minor versions from the existing version tags as we need to release fixes on top of minor versions.

Scenarion:
- `main` is the current release branch with the tag `v3.0.0`
- a new feature gets added to the `main` branch and `v3.1.0` gets released (and a git tag is created)
- another new feature gets added to the `main` branch and `v3.2.0` gets released
- A CVE comes in that we'd like to fix also fix on the previous minor versions
  - A fix for the CVE is pushed to `main` and the `v3.2.1` release is created
  - A `v3.0.x` branch is created from the `v3.0.0` tag and the commit with the fix is merged into it and `v3.0.1` is released
  - A `v3.1.x` branch is created from the `v3.1.0` tag and the commit with the fix is merged into it and `v3.0.1` is released

This way we can still use the `main` branch for the current release, and we don't need to pre-create all the release branches for each minor version since that will create a huge amount of maintenance overhead.

**Comment by Jamstah** [comment]
Are we planning for beta/rc releases? How do they work?

**Comment by milosgajdos** [comment]
> Are we planning for beta/rc releases? How do they work?

We should do. We have done that in past, but how theyre supported and how they work was never properly "codified" so we should discuss that.

**Comment by DavidSpek** [comment]
I'll need to do some testing with the automation I've described above to see how it ends up working exactly. Then I'll also play around with beta/rc releases. For `v3` I think it's obvious that this should be done. However, I don't think we'd need beta/rc releases for patch (and possibly minor) releases. I think this was done in the past mainly because the release cadence was slow which we're trying to combat here.

**Comment by wy65701436** [comment]
I appreciate the proposal, and it's important for us to establish a concrete release cadence.

- For major releases, we can aim for a 3-4 month cycle. These releases will primarily focus on introducing new features and enhancements.
- For minor/patch releases, a monthly cycle would be suitable. These releases will primarily address bugs and any identified CVEs.

Additionally, we should define a process that includes beta, RC and GA.

**Comment by Jamstah** [comment]
I think it's worth getting a beta release up ASAP even if it's not fully automated, then we can tell dependent projects about it

**Comment by milosgajdos** [comment]
Yes, we should kick things off for `v3` with an `alpha` (or `beta`) release.

Going forward I think the monthly release cycle makes sense to me 🤔

Personally, I'd prefer RCs only in scenarios that introduce some new behaviour or major b/w compatibility breakage kinda thing. Otherwise we should stick to semver releases 🤔

**Comment by thaJeztah** [comment]
I'll give this a more thorough read, but one quick comment;

> We configure the repository to only allow Squash commits 

I really prefer _NOT_ to do this. Github's squash commits are... rather bad; 

- they remove any (GPG) signing from the commits
- it's a lot more difficult to find trace back origin (e.g. the commit SHA no longer matches the commits from the PR)
- grouping commits into logical chunks of work should be part of the review process (and the author of the PR). Squashing all commits can lead to a more complicated history ("git blame" now pointing to a large diff, losing context of the changes on that specific line)

**Comment by milosgajdos** [comment]
I agree with @thaJeztah, I am also against squashing commits. We can always ask explicitly for squash on a PR if need be.

**Comment by DavidSpek** [comment]
Then we'd need to find another way to enforce conventional commits for all the commits of a PR. Which can be a pain for contributors since rewriting a commit message is less easy than changing a GitHub PR title. I'm not against it though, just need to look into how to enforce it. I hope to get something setup for this this week or early next week.

**Comment by DavidSpek** [comment]
> I appreciate the proposal, and it's important for us to establish a concrete release cadence.
> 
> * For major releases, we can aim for a 3-4 month cycle. These releases will primarily focus on introducing new features and enhancements.
> * For minor/patch releases, a monthly cycle would be suitable. These releases will primarily address bugs and any identified CVEs.
> 
> Additionally, we should define a process that includes beta, RC and GA.

@wy65701436 Major releases would only be created for breaking changes, **not** for new features and enhancements. Those are what minor releases are for. For a setup like this to work we need to be strict with enforcing SemVer, and Semantic Release helps us with enforcing SemVer strictly.

**Comment by DavidSpek** [comment]
Good suggestion. We should be able to layer that on top of conventional commits and run it in an action on PRs to check if a PR makes breaking changes.

**Comment by wy65701436** [comment]
> Major

Yes, revise my comments, should be minor as #1 and patch for #2
